File: cplusplus/Bridge_ConsoleLogger.cpp
#include <iostream>
#include "Bridge_LoggerHelpers.h"
#include "Bridge_ConsoleLogger.h"

namespace // Anonymous
{
    /// <summary>
    /// Represents a logger that throws away anything sent its way.
    /// </summary>
    class ConsoleLoggerImpl : public DesignPatternExamples_cpp::ILogger
    {
    private:

        /// <summary>
        /// Send a formatted line to the console.
        /// </summary>
        /// <param name="logLevel">The level of the log detail.</param>
        /// <param name="msg">The message to log.</param>
        void _WriteLine(const std::string& logLevel, const std::string& msg)
        {
            std::string output = DesignPatternExamples_cpp::LoggerHelpers::FormatLogLine(logLevel, msg);
            std::cout << output << std::endl;
        }

    public:
        /// <summary>
        /// Constructor.
        /// </summary>
        ConsoleLoggerImpl()
        {
        }

        // ILogger Members

        void LogTrace(const std::string& msg)
        {
            _WriteLine("TRACE", msg);
        }

        void LogInfo(const std::string& msg)
        {
            _WriteLine("INFO ", msg);
        }

        void LogError(const std::string& msg)
        {
            _WriteLine("ERROR", msg);
        }
    };

} // end anonymous namespace


namespace DesignPatternExamples_cpp
{
    /// <summary>
    /// Create an instance of a console logger, which writes to the standard
    /// output.
    /// </summary>
    /// <returns>An instance of an ILogger object.</returns>
    std::unique_ptr<ILogger> ConsoleLogger::CreateLogger()
    {
        return std::make_unique<ConsoleLoggerImpl>();
    }

} // end namespace

File: csharp/Bridge_ConsoleLogger.cs
/// @file
/// @brief
/// The @ref DesignPatternExamples_csharp.ConsoleLogger "ConsoleLogger" class
/// as used in the @ref bridge_pattern "Bridge pattern".

using System;

namespace DesignPatternExamples_csharp
{
    /// <summary>
    /// Represents a logger that writes to the console window.
    /// </summary>
    internal class ConsoleLogger : ILogger, IDisposable
    {
        private bool _disposed;

        /// <summary>
        /// Constructor.
        /// </summary>
        private ConsoleLogger()
        {
        }


        /// <summary>
        /// Send a formatted line to the console.
        /// </summary>
        /// <param name="logLevel">The level of the log detail.</param>
        /// <param name="msg">The message to log.</param>
        private void _WriteLine(string logLevel, string msg)
        {
            string output = LoggerHelpers.FormatLogLine(logLevel, msg);
            Console.WriteLine(output);
        }



        #region ILogger Members

        void ILogger.LogTrace(string msg)
        {
            _WriteLine("TRACE", msg);
        }

        void ILogger.LogInfo(string msg)
        {
            _WriteLine("INFO ", msg);
        }

        void ILogger.LogError(string msg)
        {
            _WriteLine("ERROR", msg);
        }

        #endregion

        #region IDisposable Members

        /// <summary>
        /// Shut down the logger gracefully (in this case, nothing really needs to be done).
        /// </summary>
        void IDisposable.Dispose()
        {
            if (!_disposed)
            {
                _disposed = true;
            }
        }

        #endregion


        /// <summary>
        /// Create an instance of a console logger, where logging goes to the console.
        /// </summary>
        /// <returns>An instance of an ILogger object.</returns>
        public static ILogger CreateConsoleLogger()
        {
            return new ConsoleLogger();
        }
    }
}
